### DISCUSSION TITLE
Fan - RotationSpeed - output PWM 1.2v instead of 3.3v

### DISCUSSION THREAD

**Discussion by Suitec1** [HomeSpan/HomeSpan#829] [Q&A]
Hello friends, I finally managed to create my own library for Fan PWM, but the problem is that at 100% load, only 1.2v appears on the given pin.

On the contrary, if I use a dimmable led diode with (DEV_DimmableLED) on the given pin as in the examples, a PWM signal of 3.3V normally appears at the output

Does anyone have any idea what it could be? It must be somewhere in the code.

Here is the library code.

      `struct DEV_FanPWM : Service::Fan { 
       int fanpwmPin; 
       SpanCharacteristic *power; 
       SpanCharacteristic *level; 

    DEV_FanPWM(int pin) : Service::Fan(){
        power = new Characteristic::Active(); 
        level = new Characteristic::RotationSpeed(100); 
        level->setRange(5,100,1); 
        this->fanpwmPin = pin; 
        pinMode(fanpwmPin, OUTPUT); 
        Serial.print("Configuring Fan PWM: Pin="); 
        Serial.print(fanpwmPin); 
        Serial.print("\n"); 
    } 

    boolean update(){ 
        LOG1("Updating Fan PWM on pin="); 
        LOG1(fanpwmPin); 
        LOG1(": Current power="); 
        LOG1(power->getVal()?"true":"false"); 
        LOG1(" Current RotationSpeed="); 
        LOG1(level->getVal()); 
        if(power->updated()){ 
            LOG1(" New power="); 
            LOG1(power->getNewVal()?"true":"false"); 
        } 
        if(level->updated()){ 
            LOG1(" New RotationSpeed="); 
            LOG1(level->getNewVal()); 
        } 
        LOG1("\n"); 
        analogWrite(fanpwmPin, power->getNewVal()*level->getNewVal()); 
        return(true); 
       } 
     }; `

**Comment by HomeSpan** [comment]
Are you connecting a fan directly to the ESP32 pin?  What type of fan?

**Comment by HomeSpan** [comment]
Also, why are you using analogWrite?  HomeSpan has built-in PWM that properly scales and manages the duty-cycle for you.

**Comment by Suitec1** [comment]
You're right! Otherwise, my idea is to control the fan using a Speed controller for AC motors with a PWM input for the commutator motor in the fan. 

Unfortunately, I'm not a programmer and I'm kind of learning it myself. I am an  **electrician.**

I finally managed to solve it this way.

` 


    struct DEV_PWMFAN : Service::Fan {
       LedPin *ledPin; 
       SpanCharacteristic *power;
       SpanCharacteristic *level;

     DEV_PWMFAN(int pin) : Service::Fan(){
       power=new Characteristic::Active();
       level=new Characteristic::RotationSpeed(50);
       level->setRange(1,100,1);
       this->ledPin=new LedPin(pin);

       }
     boolean update(){
       ledPin->set(power->getNewVal()*level->getNewVal());
       return(true);
       }
     };
`

**Comment by G-D-D-971** [comment]
Hi, I had a similar problem, I fixed it and I put the code I use here -> #815, so take a look, I tested myself and it works great.
PS: You can use mosfet to serve as a power relay, it costs less and it is very space-saving...
